### DISCUSSION TITLE
Why doesn't WebGL2 queryCounterEXT link?

### DISCUSSION THREAD

**Discussion by Joyoslive** [emscripten-core/emscripten#19131] [Q&A]
Hi,
I've been trying to profile a webgl2 renderer that compiles with emscripten. I found the header file containing all the necessary defines as well as a prototype function for emscripten_webgl2_queryCounterEXT. However, when I try to use this function I get an "undefined symbol" error. If anyone has any idea as to what I need to do to make this link correctly I would be very grateful.

**Comment by sbc100** [comment]
I believe the indended user-facing symbol is `glQueryCounterEXT` .  This seems to be defined in library_webgl.js: https://github.com/emscripten-core/emscripten/blob/574f4f449780f63ab97970d5d6b5625e124c7d79/src/library_webgl.js#L1848-L1854

Does that symbol not work for you?

**Reply by Joyoslive** [reply]
glQueryCounterEXT does not seem to exist in C++. Where would this be declared? I might have missed including some header.

**Reply by sbc100** [reply]
I see a declaration in `GLES2/gl2ext.h` and in `webgl/webgl1_ext.h`.. are they don't working for some reason?:

```
$ git grep glQueryCounterEXT system/include/
system/include/GLES2/gl2ext.h:GL_APICALL void GL_APIENTRY glQueryCounterEXT (GLuint id, GLenum target);
system/include/webgl/webgl1_ext.h:WEBGL_APICALL void GL_APIENTRY emscripten_glQueryCounterEXT(GLuint id, GLenum target);
system/include/webgl/webgl1_ext.h:WEBGL_APICALL void GL_APIENTRY glQueryCounterEXT(GLuint id, GLenum target);
```

**Comment by sbc100** [comment]
It looks like `emscripten_webgl2_queryCounterEXT` was added to system/lib/gl/webgl2_ext.h back in #6254 but I don't see anywhere where that symbol is defined, even back in that change.  @juj, was this a mistake?  It doesn't seem the follow the naming convention of any other symbols.

**Comment by juj** [comment]
> I believe the indended user-facing symbol is glQueryCounterEXT

This is correct.

timer queries in WebGL are little bit awkward, because there are two separate extension versions depending on which core WebGL context version one initializes. In Emscripten we actually paper over this behavior, and fuse the functions into one.

> It looks like emscripten_webgl2_queryCounterEXT was added to system/lib/gl/webgl2_ext.h back in https://github.com/emscripten-core/emscripten/pull/6254 but I don't see anywhere where that symbol is defined, even back in that change. @juj, was this a mistake?

This indeed is a mistake.

> It doesn't seem the follow the naming convention of any other symbols.

The naming convention here is a bit complex, and deserves a mention.

For all WebGL extensions that define new functions, the function naming of the functions overlays on top of the existing GLES2/GLES3 function names for portability.

So for example, [ANGLE_instanced_arrays](https://registry.khronos.org/webgl/extensions/ANGLE_instanced_arrays/) extension provides three functions

```
    undefined drawArraysInstancedANGLE(GLenum mode, GLint first, GLsizei count, GLsizei primcount);
    undefined drawElementsInstancedANGLE(GLenum mode, GLsizei count, GLenum type, GLintptr offset, GLsizei primcount);
    undefined vertexAttribDivisorANGLE(GLuint index, GLuint divisor); 
```
In the native GLES2/GLES3 world, the same types of extensions existed, with native `gl*` function API names. So these functions became

```
WEBGL_APICALL void GL_APIENTRY emscripten_glDrawArraysInstancedANGLE(GLenum mode, GLint first, GLsizei count, GLsizei primcount);
WEBGL_APICALL void GL_APIENTRY emscripten_glDrawElementsInstancedANGLE(GLenum mode, GLsizei count, GLenum type, GLintptr offset, GLsizei primcount);
WEBGL_APICALL void GL_APIENTRY emscripten_glVertexAttribDivisorANGLE(GLuint index, GLuint divisor);
WEBGL_APICALL void GL_APIENTRY glDrawArraysInstancedANGLE(GLenum mode, GLint first, GLsizei count, GLsizei primcount);
WEBGL_APICALL void GL_APIENTRY glDrawElementsInstancedANGLE(GLenum mode, GLsizei count, GLenum type, GLintptr offset, GLsizei primcount);
WEBGL_APICALL void GL_APIENTRY glVertexAttribDivisorANGLE(GLuint index, GLuint divisor);
```
(The `emscripten_` aliases being there for the `glGetProcAddress()` and offscreen framebuffer features)

However, there are then some extensions that are unique to WebGL that there does not exist a corresponding native GLES2/GLES3 extension.

Currently such extensions are:
 - [WEBGL_lose_context](https://www.khronos.org/registry/webgl/extensions/WEBGL_lose_context/)
 - [WEBGL_debug_shaders](https://www.khronos.org/registry/webgl/extensions/WEBGL_debug_shaders/)
 - [WEBGL_compressed_texture_astc](https://www.khronos.org/registry/webgl/extensions/WEBGL_compressed_texture_astc/)

These come with new WebGL functions that don't exist in native GL/GLES. For example:

```
[Exposed=(Window,Worker), LegacyNoInterfaceObject]
interface WEBGL_lose_context {
      undefined loseContext();
      undefined restoreContext();
};
```

 I did not want to step in to the existing native GL/GLES `gl*` namespace and invent new native GL sounding function names, such as

```
glLoseContext();
glRestoreContext();
```

since that would kind of confuse with the existing native specifications and make it read like it was an official core GL/GLES thing.

So for these functions, I opted to go for the naming scheme

```
emscripten_webgl_....()
emscripten_webgl2_....()
```
for WebGL 1 and WebGL 2 extensions and these extensions will be available by a `#include <webgl/webgl1_ext.h>` and `#include <webgl/webgl2_ext.h>` respectively.

The [EXT_disjoint_timer_query_webgl2](https://registry.khronos.org/webgl/extensions/EXT_disjoint_timer_query_webgl2/) extension brings a new function, so that is probably why I had written a declaration of a function `emscripten_webgl2_queryCounterEXT` in webgl2_ext.h. 

However, the implementation of that function was actually merged into the general `glQueryCounterEXT` function, so no new functions were actually created. Posted PR #19158 to clean up this function name.
